infrastructure/kctf/base/nsjail-docker/files/proof_of_work/pow.py

```diff
@@ -1,124 +1,125 @@
 #!/usr/bin/env python3
 # -*- coding: utf-8 -*-
 
-import hmac
+import base64
 import secrets
-import sys
 import socket
+import sys
+
+VERSION = 's'
+MODULUS = 2**1279-1
+CHALSIZE = 2**128
 
 SOLVER_URL = 'https://github.com/google/google-ctf/blob/master/infrastructure/kctf/base/nsjail-docker/proof_of_work/pow.py'
 
 def stdin_is_localhost_socket():
     try:
         sock = socket.socket(fileno=sys.stdin.fileno())
+        peername = sock.getpeername()
+        sock.detach()
+        return peername[0] == '::ffff:127.0.0.1'
     except OSError:
         return False
 
-    peername = sock.getpeername()
-    sock.detach()
-
-    return peername[0] == '::ffff:127.0.0.1'
-
-def gen_seed():
-    return secrets.token_hex(8)
+def sloth_root(x, diff, p):
+    for i in range(diff):
+        x = pow(x, (p + 1) // 4, p) ^ 1
+    return x
 
-def byte_has_leading_zeroes(b, count):
-    if count == 0:
-        return True
+def sloth_square(y, diff, p):
+    for i in range(diff):
```

Comment:
What am I missing here, finding the solution and verifying it is both a loop over range(diff). How is verifying faster?

Reply:
exponentiation of squares is very fast, compared to exponentiation to a large exponent